cosmos/lib/pagination.go

```diff
@@ -0,0 +1,37 @@
+package lib
+
+import (
+	"github.com/cosmos/cosmos-sdk/types/query"
+)
+
+type PageResponse interface {
+	GetPagination() *query.PageResponse
+}
+
+func getNextPaginationKey(response PageResponse) []byte {
+	if response == nil {
+		return nil
+	}
+	pagination := response.GetPagination()
+	if pagination == nil {
+		return nil
+	}
+	return pagination.NextKey
+}
+
+// Paginate returns the aggregate responses of a paginated cosmos query
+func Paginate[T PageResponse](
+	responseFetcher func(paginationKey []byte) (T, error),
+) (responses []T, err error) {
+	var response T
+
+	for paginationKey, i := []byte{}, 0; len(paginationKey) > 0 || i == 0; paginationKey, i = getNextPaginationKey(response), i+1 {
+		response, err = responseFetcher(paginationKey)
+		if err != nil {
+			return nil, err
+		}
+		responses = append(responses, response)
+	}
+
+	return responses, nil
+}
```

Comment:
We can add another function argument here to reduce the boilerplate of parsing results from responses (e.g. `response.GetSupply`), but it seems to reduce readability for very little gain

```go
// Paginate returns the aggregate result of a paginated cosmos query
func Paginate[T PageResponse, U interface{}](
	responseFetcher func(paginationKey []byte) (T, error),
	resultGetter func(response T) []U,
) ([]U, error) {
	var result []U
	var response T
	var err error

	for paginationKey, i := []byte{}, 0; len(paginationKey) > 0 || i == 0; paginationKey, i = getNextPaginationKey(response), i+1 {
		response, err = responseFetcher(paginationKey)
		if err != nil {
			return nil, err
		}
		result = append(result, resultGetter(response)...)
	}

	return result, nil
}
```

```go
supply, err := cosmlib.Paginate[*banktypes.QueryTotalSupplyResponse, types.Coin](
	func(paginationKey []byte) (*banktypes.QueryTotalSupplyResponse, error) {
		return c.querier.TotalSupply(ctx, &banktypes.QueryTotalSupplyRequest{
			Pagination: &query.PageRequest{Key: paginationKey},
		})
	},
	func(response *banktypes.QueryTotalSupplyResponse) []types.Coin {
		return response.GetSupply()
	},
)
```